### PR TITLE
[api-minor] Don't normalize the text used in the text layer.

### PR DESCRIPTION
**Pull request by calixteman**
Some arabic chars like \ufe94 could be searched in a pdf, hence it must be normalized when creating the search query. So to avoid to duplicate the normalization code, everything is moved in the find controller.
The previous code to normalize text was using NFKC but with a hardcoded map, hence it has been replaced by the use of normalize("NFKC") (it helps to reduce the bundle size by 30kb).
In playing with this \ufe94 char, I noticed that the bidi algorithm wasn't taking into account some RTL unicode ranges, the generated font wasn't embedding the mapping this char and the unicode ranges in the OS/2 table weren't up-to-date.

When normalized some chars can be replaced by several ones and it induced to have some extra chars in the text layer. To avoid any regression, when copying some text from the text layer, a copied string is normalized (NFKC) before being put in the clipboard (it works like this in either Acrobat or Chrome).

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
Given that we should be pretty close to the next PDF.js release, this is probably not something that we want to land just prior.
In the mean-time, I've got a couple of questions and comments based on a quick look:
 - The bundle-size decrease is *really* nice :-)
 
 - This is probably too much of an API-break as-is, and I do believe that we'll need to maintain the old behaviour (which has existed since "forever") by normalizing by *default*.
   So can we please add a new `getTextContent`/`streamTextContent` parameter, e.g. named `disableNormalization = false` and thus off by default, that causes the worker-thread to call `normalize("NFKC")` on the text-content it returns?
   The relevant viewer call-sites, and the browsertest `Driver`, would then simply call these API-methods with `disableNormalization = true` set to get the new behaviour.
 
 - How does text-selection work in documents that mix LTR and RTL text, especially on the same line?
    Previously we'd reverse e.g. arabic text inside of text-runs that *mostly* contains LTR text, although I suppose that probably never worked perfectly.

 - How does this work with a11y-software?
   Will it be able to "make sense" of any ligatures found in the textLayer DOM, since they were previously always available in their "expanded" form in the DOM?

 - Does this work in Safari?
   *Please note:* I don't mean that attempting to support Safari should in any way block or complicate things here, however if this doesn't work in Safari we might simply need to update (or possibly even remove) its support status from the FAQ.

**Comment by Snuffleupagus** [comment]
> Good question, I tried with a fi ligature and NVDA and it doesn't work correctly:
> [nvaccess/nvda#14740](https://github.com/nvaccess/nvda/issues/14740)

The discussion in that issue seem to suggest that the actual bug lies elsewhere; what's the overall status of a11y support for ligatures?

> The previous code to normalize text was using NFKC but with a hardcoded map, hence it has been replaced by the use of normalize("NFKC") (it helps to reduce the bundle size by 30kb).

This actually seems problematic as-is, since we'll now end up normalizing *a lot* more than before. Take e.g. the [fraction-highlight.pdf](https://github.com/mozilla/pdf.js/blob/master/test/pdfs/fraction-highlight.pdf) document as an example, where copy-and-paste currently gives the following result:
```
some text before a fraction
½
some text after a fraction
½¼¾
more fractions!
```
With this patch you'll instead get:
```
some text before a fraction
1⁄2
some text after a fraction
1⁄21⁄43⁄4
more fractions!
```

This seems problematic for a couple of reasons:
 - We'd then have different copy-and-paste behaviour compared to both Adobe Reader *and* PDFium.
 - Normalizing those kind of combined-characters feels, in my opinion, to be neither particularly helpful nor what the user likely wants here.
 - It's not clear how/if this works with searching, if you copy one of those "expanded"-characters and then use that as input when searching?

To me this feels like opening the door to loads of bugs both now and down the line, by us essentially relinquishing any control over "when" normalization happens. To that end I wonder if we could perhaps combine the current hard-coded map with using standard normalization (which should still be a reduction in code-size)?
```javascript
const NORMALIZE_UNICODES = new Set([
  "\u00A8",
  "\u00AF",
  // The rest of the "keys" from `getNormalizedUnicodes` goes here...
]);

function normalizeTextContent(str) {
  const buf = [];
  for (const char of str) {
    buf.push(NORMALIZE_UNICODES.has(char) ? char.normalize("NFKC") : char);
  }
  return buf.join("");
```
You'd then call that helper-function on copy-and-paste, and we'd still have control of normalization without having to hard-code everything like previously.

**Comment by calixteman** [comment]
All the chars we've in `getNormalizedUnicodes` aren't normalized when they're copied in Acrobat.
For example `\u0132` is normalized when a string is searched but it isn't when it's copied.

File: test/unit/unicode_spec.js [hunk 1]
```diff
@@ -152,69 +150,8 @@ describe("unicode", function () {
       expect(getUnicodeRangeFor(0x0041)).toEqual(0);
       // fi (Alphabetic Presentation Forms)
       expect(getUnicodeRangeFor(0xfb01)).toEqual(62);
-    });
-
-    it("should not get a Unicode range", function () {
-      expect(getUnicodeRangeFor(0x05ff)).toEqual(-1);
-    });
```
Comment (Snuffleupagus) [inline test/unit/unicode_spec.js:159]: Wouldn't it make sense to keep a test-case for the fallback code-path where `-1` is returned?

File: test/unit/api_spec.js [hunk 2]
```diff
@@ -2475,20 +2487,39 @@ page 1 / 3`);
...
-      expect(
-        text.includes(`Open Sans is a humanist sans serif typeface designed by Steve Matteson.
+      expect(text.includes(expected)).toEqual(true);
+
+      ({ items } = await pdfPage.getTextContent({
+        disableNormalization: false,
+      }));
+      text = mergeText(items);
+
+      expect(text.includes(expected)).toEqual(false);
```
Comment (Snuffleupagus) [inline test/unit/api_spec.js:2511]: This line is probably no longer necessary, with the explicit check now being used below.

File: src/shared/util.js [hunk 3]
```diff
@@ -1026,6 +1026,16 @@ function createPromiseCapability() {
...
+function getNormalizeRegex() {
+  // In order to generate the following regex:
+  //  - create a PDF containing all the chars in the range 0000-FFFF with
+  //    a NFKC which is different of the char.
+  //  - copy and paste all those chars and get the ones where NFKC is
+  //    required.
+  // It appears that most the chars here contain some ligatures.
+  return /[\u00a0\u00b5\u037e\u0eb3\u2000-\u200a\u202f\u2126\ufb00-\ufb04\ufb06\ufb20-\ufb36\ufb38-\ufb3c\ufb3e\ufb40-\ufb41\ufb43-\ufb44\ufb46-\ufba1\ufba4-\ufba9\ufbae-\ufbb1\ufbd3-\ufbdc\ufbde-\ufbe7\ufbea-\ufbf8\ufbfc-\ufbfd\ufc00-\ufc5d\ufc64-\ufcf1\ufcf5-\ufd3d\ufd88\ufdf4\ufdfa-\ufdfb\ufe71\ufe77\ufe79\ufe7b\ufe7d]+/gu;
+}
+
```
Comment (Snuffleupagus) [inline src/shared/util.js:1038]: Would it perhaps make sense to just expose a helper function that does the normalization instead, rather than only the regular expressions, e.g. something like the following since that'd slightly simplify both call-sites?
```suggestion
let NormalizeRegex;

function normalizeUnicode(str) {
  if (!NormalizeRegex) {
    // In order to generate the following regex:
    //  - create a PDF containing all the chars in the range 0000-FFFF with
    //    a NFKC which is different of the char.
    //  - copy and paste all those chars and get the ones where NFKC is
    //    required.
    // It appears that most the chars here contain some ligatures.
    NormalizeRegex = /[\u00a0\u00b5\u037e\u0eb3\u2000-\u200a\u202f\u2126\ufb00-\ufb04\ufb06\ufb20-\ufb36\ufb38-\ufb3c\ufb3e\ufb40-\ufb41\ufb43-\ufb44\ufb46-\ufba1\ufba4-\ufba9\ufbae-\ufbb1\ufbd3-\ufbdc\ufbde-\ufbe7\ufbea-\ufbf8\ufbfc-\ufbfd\ufc00-\ufc5d\ufc64-\ufcf1\ufcf5-\ufd3d\ufd88\ufdf4\ufdfa-\ufdfb\ufe71\ufe77\ufe79\ufe7b\ufe7d]+/gu;
  }
  return str.replaceAll(NormalizeRegex, m => m[0].normalize("NFKC"));
}

```

**Review by Snuffleupagus — COMMENTED** [review]
Would it be feasible to add an integration-test, using a PDF document with ligatures, that checks that the copied text is actually being normalized as expected?

File: web/text_layer_builder.js [hunk 4]
```diff
@@ -212,6 +221,19 @@ class TextLayerBuilder {
...
+
+    div.addEventListener("copy", event => {
+      const selection = document.getSelection();
+      event.clipboardData.setData(
+        "text/plain",
+        selection
+          .toString()
+          .replaceAll(TextLayerBuilder.NormalizeRegex, m =>
+            m[0].normalize("NFKC")
+          )
```
Comment (Snuffleupagus) [inline web/text_layer_builder.js:233]: Something that I just realized is that we need to remove null chars here as well, since copying otherwise won't work correctly in all cases; note what happens when selecting and copying the contents of "Table 3.15" from this page: https://ftp.acc.umu.se/mirror/CTAN/info/lshort/english/lshort.pdf#page=83
```suggestion
        removeNullCharacters(
          selection
            .toString()
            .replaceAll(TextLayerBuilder.NormalizeRegex, m =>
              m[0].normalize("NFKC")
            )
        )
```

File: web/pdf_find_controller.js [hunk 5]
```diff
@@ -126,12 +126,35 @@ function normalize(text) {
...
+                range[1]
+              )}`
+            );
+          }
+          range[0] = range[1] = i;
+        } else {
+          range[1] = i;
+        }
+      }
+    }
```
Comment (Snuffleupagus) [inline web/pdf_find_controller.js:155]: How does all of this "extra" parsing here affect performance of the initial search operation?

Comment (calixteman) [reply web/pdf_find_controller.js:155]: On my computer it takes around 7,8 ms (I added some `performance.mark` around that code and then measure with the profiler).
The final string has a length of 315 chars.
We can hard-code the string but if there are some changes in normalization stuff (on the js engine side) we would have to regenerate that string, but maybe it's something which will never happen, I don't know.
So if you think it's too slow, then we hard-code the string and we wait for complaints.

Comment (Snuffleupagus) [reply web/pdf_find_controller.js:155]: > On my computer it takes around 7,8 ms (I added some `performance.mark` around that code and then measure with the profiler).

That doesn't look too bad on its own, although I guess this is on powerful hardware.
However possibly a better question, on my part, would've been how much time *in percentage* of the initial `normalize`-call is spent in this new code-block?

> We can hard-code the string but if there are some changes in normalization stuff (on the js engine side) we would have to regenerate that string, but maybe it's something which will never happen, I don't know.

That's indeed a risk if we just hard-code this data, although I'd be surprised if this changes a lot.

> So if you think it's too slow, then we hard-code the string and we wait for complaints.

How about a compromise? What if we hard-code the data, but let the existing find unit-tests check that normalization hasn't changed by instead moving this logic into a helper function in `pdf_find_utils.js`?
```diff
diff --git a/web/pdf_find_utils.js b/web/pdf_find_utils.js
index 24ec4c575..60964973b 100644
--- a/web/pdf_find_utils.js
+++ b/web/pdf_find_utils.js
@@ -112,4 +112,47 @@ function getCharacterType(charCode) {
   return CharacterType.ALPHA_LETTER;
 }
 
-export { CharacterType, getCharacterType };
+let NormalizeWithNFKC;
+
+function getNormalizeWithNFKC() {
+  NormalizeWithNFKC ||=
+    "¨ª¯²²-µ¸-º¼-¾Ĳ-ĳĿ-ŀſǄǄ-ǌǱ-ǳʰ-ʸ˘-˝ˠ-ˤͺ;΄΄-΅ϐϐ-ϖϰ-ϲϴ-ϵևٵٵ-ٸक़-य़ড়-ঢ়ਲ਼ਸ਼ਖ਼ਖ਼-ਜ਼ଡ଼ଡ଼-ଢ଼ຳໜໜ-ໝགྷཌྷདྷབྷཛྷཀྵჼᴬᴬ-ᴮᴰ-ᴺᴼ-ᵍᵏ-ᵪᶛᶛ-ᶿẚ-ẛέήίόύώΆ᾽᾽-῁Ή῍῍-῏Ί῝῝-῟Ύ῭῭-`Ώ´´-῾ - ‗․․-…″″-‴‶-‷‾⁇⁇-⁉ ⁰⁰-ⁱ⁴-₎ₐ-ₜ℀℀-℃℅-ℇ℉-ℓℕ-№ℙ-ℝ℠-™ΩℨKK-ℭℯ-ℱℳ-ℹ℻-⅀ⅅ-ⅉ⅐-ⅿ∬∬-∭∯-∰〈-〉①-⓪⩴⩴-⩶ⱼⱼ-ⱽ⺟⻳⼀⼀-⿕〶〸〸-〺゛-゜ヿㄱㄱ-ㆎ㆒-㆟㈀-㈞㈠-㉇㉐-㉾㊀-㏿ꚜ-ꚝꟲꟲ-ꟴꟸ-ꟹꭜ-ꭟ豈豈-嗀晴凞凞-羽諸逸逸-都飯-舘並-龎ﬀ-ﬆﬓ-ﬗײַײַ-זּטּ-לּנּנּ-סּףּ-פּצּ-ﮱﯓ-ﴽﵐ-ﶏﶒ-ﷇﷰ-﷼︐-︙︰-﹄﹇-﹒﹔-﹦﹨-﹫ﹰ-ﹲﹶﹶ-ﻼ！-ﾾￂ-ￇￊ-ￏￒ-ￗￚ-ￜ￠-￦";
+
+  if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
+    const ranges = [];
+    const range = [];
+    const diacriticsRegex = /^\p{M}$/u;
+    // Some chars must be replaced by their NFKC counterpart during a search.
+    for (let i = 0; i < 65536; i++) {
+      const c = String.fromCharCode(i);
+      if (c.normalize("NFKC") !== c && !diacriticsRegex.test(c)) {
+        if (range.length !== 2) {
+          range[0] = range[1] = i;
+          continue;
+        }
+        if (range[1] + 1 !== i) {
+          if (range[0] === range[1]) {
+            ranges.push(c);
+          } else {
+            ranges.push(
+              `${String.fromCharCode(range[0])}-${String.fromCharCode(
+                range[1]
+              )}`
+            );
+          }
+          range[0] = range[1] = i;
+        } else {
+          range[1] = i;
+        }
+      }
+    }
+    if (ranges.join("") !== NormalizeWithNFKC) {
+      throw new Error(
+        "getNormalizeWithNFKC - update the `NormalizeWithNFKC` string."
+      );
+    }
+  }
+  return NormalizeWithNFKC;
+}
+
+export { CharacterType, getCharacterType, getNormalizeWithNFKC };
```

File: test/integration/copy_paste_spec.js [hunk 6]
```diff
@@ -117,4 +117,47 @@ describe("Copy and paste", () => {
...
+          );
+
+          const text = await page.evaluate(() =>
+            navigator.clipboard.readText()
+          );
+
+          expect(!!text).withContext(`In ${browserName}`).toEqual(true);
+          expect(text)
+            .withContext(`In ${browserName}`)
+            .toEqual("abcdeffffiflffifflﬅstghijklmno");
```
Comment (Snuffleupagus) [inline test/integration/copy_paste_spec.js:158]: It looks like there's one non-normalized ligature, ie. `ﬅ` one, left here?

Comment (Snuffleupagus) [reply test/integration/copy_paste_spec.js:158]: In the `master` branch this copies as `"abcdeffffiflffifflſtstghijklmno"`.

Comment (calixteman) [reply test/integration/copy_paste_spec.js:158]: Yes I know: it works like that in Acrobat.
In Chrome and Foxit, this char is mapped on `ſt` (the first char is a `\u017F`, e.g. long 's') and NFKC changes it to `st`.
We can add it in the list of the char to normalize and then we'll get a `st`.
I'm not an expert in latin, hence I don't know what a user could expect to have.

Comment (Snuffleupagus) [reply test/integration/copy_paste_spec.js:158]: It's annoying that there's no consensus between PDF Viewers here :-(

Given that we previously normalized this and that Chrome also does so (since nowadays it feels like users often compare with Chrome), I'm leaning towards suggesting that we should keep normalizing that ligature.

File: web/pdf_viewer.js [hunk 7]
```diff
@@ -655,6 +655,9 @@ class PDFViewer {
   async getAllText() {
     const texts = [];
     const buffer = [];
+    const options = {
+      disableNormalization: false,
+    };
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:660]: Do we actually need this, since that's the default value and we've unit-tests that'll fail if we ever change the default value in the API?

File: web/pdf_viewer.js [hunk 8]
```diff
@@ -758,6 +761,7 @@ class PDFViewer {
     if (!pdfDocument) {
       return;
     }
+
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:764]: Nit: unrelated change.

**Review by Snuffleupagus — COMMENTED** [review]
With the latest round of comments addressed, let's start running tests to see what the "fallout" looks like :-)

File: web/pdf_viewer.js (hunk 7 above)
Comment (calixteman) [reply web/pdf_viewer.js:660]: My idea was to add it just for a matter of documentation, but I can add a comment.

**Comment by Snuffleupagus** [comment]
From a *brief* look, the textLayer movement seem fine and expected as far as I can tell.

The movement in regular `eq`-tests on Linux seem strange, however it actually looks like slight improvements to me.

/botio integrationtest

**Comment by calixteman** [comment]
I don't see anything really wrong

> From a _brief_ look, the textLayer movement seem fine and expected as far as I can tell.
> 
> The movement in regular `eq`-tests on Linux seem strange, however it actually looks like slight improvements to me.
> 

I'd say that the root cause is the one mentioned by Jonathan:
https://github.com/mozilla/pdf.js/pull/15157#discussion_r917782950
this patch is modifying how we check if a char is in the private area:
https://github.com/mozilla/pdf.js/pull/16200/files#diff-d6af99a911b977730586b335a3c7ee702a383cac240414a011ab5534cda1ff3aR490-R492
where before:
https://github.com/mozilla/pdf.js/pull/16200/files#diff-d6af99a911b977730586b335a3c7ee702a383cac240414a011ab5534cda1ff3aL544
it was wrong.
Consequently, for `ﬁ` (aka, `\uFB01`) this patches changes the value we've in the extra data we add in the font and then Jonathan's arrives.

**Review by Snuffleupagus — APPROVED** [review]
r=me, thank you!

Once landed, can you please update PDF.js in mozilla-central so that we can get a little bit of real-world testing done before the next PDF.js library release?
